vmdb/db/migrate/20150405141637_remove_port_config_from_container_service.rb

```diff
@@ -0,0 +1,54 @@
+class RemovePortConfigFromContainerService < ActiveRecord::Migration
+  def up
+    create_table :container_service_port_configs do |t|
+      t.string     :ems_ref
+      t.string     :name
+      t.integer    :port
+      t.string     :target_port
+      t.string     :protocol
+      t.belongs_to :container_service, :type => :bigint
+    end
+
+    copy_over_configs
+    remove_columns
+  end
+
+  def copy_over_configs
+    ContainerService.find_each do |service|
+      ContainerServicePortConfig.create(
+          :ems_ref              => "#{service[:ems_ref]}_#{service[:port]}_"\
+                                   "#{service[:container_port]}_#{service[:protocol]}",
+          :port                 => service.port,
+          :protocol             => service.protocol,
+          :target_port          => service.container_port,
+          :container_service_id => service.id
+      )
+    end
+  end
+
+  def add_columns
+    add_column :container_services, :port, :integer
+    add_column :container_services, :protocol, :string
+    add_column :container_services, :container_port, :integer
+  end
+
+  def remove_columns
+    remove_column :container_services, :port
+    remove_column :container_services, :protocol
+    remove_column :container_services, :container_port
+  end
+
+  def down
+    add_columns
+
+    ContainerServicePortConfig.find_each do |port_config|
+      s = ContainerService.where(:id => port_config[:container_service_id]).first # There should be only one
```

Comment:
Doing this query in a loop can be expensive.  You can do this outside the loop once and use `index_by` to put it into a Hash: `indexed_containers = ContainerService.all.index_by(&:id)`.  Then that Hash can be used in the loop withioutt the N+1 penalty: `s = indexed_containers[port_config[:container_service_id]]`